src/software/thunderscope/binary_context_managers.py

```diff
@@ -528,19 +540,283 @@ def send_ci_input(
+        self.current_timestamp = int(time.time_ns())
+        self.show_gui = show_gui
+
+    def __enter__(self):
+        if not os.path.exists("/opt/tbotspython/autoReferee/bin/autoReferee"):
+            logging.info(
+                "Could not find autoref binary, did you run ./setup_software.sh"
+            )
 
-        # Send a request to the host with the size of the message
-        self.ci_socket.send(
-            encoder._VarintBytes(size) + ci_ci_input.SerializeToString()
+        self.auto_ref_proc_thread = threading.Thread(
+            target=self.start_autoref, daemon=True
         )
+        self.auto_ref_proc_thread.start()
 
-        response_data = self.ci_socket.recv(
-            Gamecontroller.CI_MODE_OUTPUT_RECEIVE_BUFFER_SIZE
+        self.auto_ref_wrapper_thread = threading.Thread(
+            target=self.send_to_autoref_and_forward_to_gamecontroller
         )
+        self.auto_ref_wrapper_thread.start()
+
+        return self
+
+    def send_geometry(self):
+        """
+        Sends updated field geometry to the AutoRef so that the TigersAutoref knows about field sizes.
+        """
+        ssl_wrapper = self.wrapper_buffer.get(block=True)
+        ci_input = AutoRefCiInput()
+        ci_input.detection.append(ssl_wrapper.detection)
+
+        field = tbots.Field.createSSLDivisionBField()
+        ci_input.geometry.CopyFrom(createGeometryData(field, 0.3))
+
+        self.ci_socket.send(ci_input)
+
+        response_data = None
+        while True:
+            try:
+                response_data = self.ci_socket.receive(AutoRefCiOutput)
+                break
+            except SslSocketProtoParseException as parse_err:
+                logging.info("error with sending geometry data: \n" + parse_err.args)
+
+        for ci_output in response_data:
+            self.forward_to_gamecontroller(ci_output.tracker_wrapper_packet)
+
+    def persistently_connect_to_autoref(self):
+        """
+        Connect to the TigersAutoref binary. Retry connection a few times if the connection doesn't go through in case the binary hasn't started yet.
+
+        :return: True if the action was successful, False otherwise
+        """
+        tries = 0
+        while tries < TigersAutoref.AUTOREF_NUM_RETRIES:
+            try:
+                # We cannot start the Autoref binary immediately, so we must wait until the binary has started before we try to connect to it
+                time.sleep(0.5)
+                self.ci_socket = SslSocket(TigersAutoref.AUTOREF_COMM_PORT)
+                return True
+            except ConnectionRefusedError:
+                tries += 1
+
+        return False
+
+    def send_to_autoref_and_forward_to_gamecontroller(self):
+        """
+        Main communication loop that sets up the TigersAutoref and coordinates communication between Simulator, TigersAutoref and Gamecontroller. Returns early if connection to the TigersAutoref binary was unsuccessful.
```

Comment:
Can we shorten the number of characters per line in this file as well?

Reply:
done